merkle-tree/indexed/tests/tests.rs

```diff
@@ -536,3 +542,251 @@ pub fn functional_non_inclusion_test() {
         .verify_non_inclusion_proof(&non_inclusion_proof)
         .unwrap();
 }
+
+#[derive(Debug, Default, Clone, Copy, BorshSerialize, BorshDeserialize)]
+pub struct RawIndexedElement {
+    pub value: [u8; 32],
+    pub next_index: usize,
+    pub next_value: [u8; 32],
+    pub index: usize,
+    pub next_value_index: usize,
+    pub next_value_next_value: [u8; 32],
+    pub next_value_path: [[u8; 32]; 10],
+}
+
+/// Check whether low element has changed
+/// - possibilities:
+///   1. check Merkle proof of -> too expensive costs like 20k CU
+///   2. check whether the array contains the index of the low element if yes then check the next value
+///   3. insert the index in to hash set that stores key: index value: next_value, can be overwritten if capacity requires,
+///         the hashset needs to be able to hold the same element multiple times though, does it?
+///         maybe not, maybe we can just say that we allow every element once and when we use it then we overwrite it immediately instead of overwriting an expired element
+///  
+pub fn patch_low_element<const CHANGELOG_CAPACITY: usize>(
+    value: &BigUint,
+    low_element: &IndexedElement<usize>,
+    changelog_array: &[RawIndexedElement; CHANGELOG_CAPACITY],
+) -> Option<(IndexedElement<usize>, [u8; 32], Option<[[u8; 32]; 10]>)> {
+    // TODO: bench how find performs vs filter
+    // filter would have the advantage that it allows for multiple changes in between
+    // although we just need most recent, then we can zero out the value that we read
+    let changelog_element_index = changelog_array
+        .iter()
+        .position(|element| element.index == low_element.index);
```

Comment:
And with this approach, I think we will get only the **oldest** change? `.position()` is going to return the first, not last matching element.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0b4d1e6c2395acdc8e665cebfdea459b

In case there were multiple updates of the same low element, that might produce bad results.

Let's think of an example.

We start with ranges `[0, 20, 60]`.

We want to insert `40`.

In the meantime, someone inserted `50` and `30`. Which produces the following changelog entries:

* `20 -> 50`
* `50 -> 60`

* `20 -> 30`
* `30 -> 40`

If we use `position()` without reverse iteration, we will get the `20 -> 50` event only. Which will make us think that we have to patch only next_index. We are going to miss the fact that `20 -> 30` change swapped the low element itself and tells us to patch the proof.

Reply:
I think a quick fix would be just reversing the iterator `.iter().rev()`.

Reply:
true, I think I will just zero out the element when I use it.
Then it cannot be found in the future.